diff --git a/dlls/dbghelp/dbghelp.c b/dlls/dbghelp/dbghelp.c
index 911daec..20c1fd5 100644
--- a/dlls/dbghelp/dbghelp.c
+++ b/dlls/dbghelp/dbghelp.c
@@ -458,6 +458,11 @@ BOOL WINAPI SymInitializeW(HANDLE hProcess, PCWSTR UserSearchPath, BOOL fInvadeP
 
     TRACE("(%p %s %u)\n", hProcess, debugstr_w(UserSearchPath), fInvadeProcess);
 
+    if (!hProcess)
+    {
+        WARN("NULL process handle in SymInitializeW, returning FALSE\n");
+        return FALSE;
+    }
     if (process_find_by_handle(hProcess))
     {
         WARN("the symbols for this process have already been initialized!\n");
@@ -548,6 +553,11 @@ BOOL WINAPI SymCleanup(HANDLE hProcess)
 
     TRACE("(%p)\n", hProcess);
 
+    if (!hProcess)
+    {
+        WARN("NULL process handle in SymCleanup\n");
+        return FALSE;
+    }
     for (ppcs = &process_first; *ppcs; ppcs = &(*ppcs)->next)
     {
         if ((*ppcs)->handle == hProcess)
