diff --git a/dlls/imm32/ime.c b/dlls/imm32/ime.c
index b407de2..a6ef653 100644
--- a/dlls/imm32/ime.c
+++ b/dlls/imm32/ime.c
@@ -464,6 +464,7 @@ static LRESULT WINAPI ime_ui_window_proc( HWND hwnd, UINT msg, WPARAM wparam, LP
     case WM_IME_CONTROL:
         FIXME( "hwnd %p, himc %p, msg %s, wparam %s, lparam %#Ix stub!\n", hwnd, himc,
                debugstr_wm_ime(msg), debugstr_imc(wparam), lparam );
+        if (wparam == IMC_CLOSESTATUSWINDOW) return 0;
         return 1;
     }
 
